File: Pharmaceutical_Company_Example/Structural/Python/Composite.py
# Component
class RegulatoryProcess:
    def execute(self):
        raise NotImplementedError("This method should be implemented by subclasses.")


# Leaf
class Step(RegulatoryProcess):
    def __init__(self, step_name):
        self.step_name = step_name

    def execute(self):
        print(f"Executing step: {self.step_name}")


# Composite
class CompositeProcess(RegulatoryProcess):
    def __init__(self, process_name):
        self.process_name = process_name
        self.children = []

    def add(self, regulatory_process):
        self.children.append(regulatory_process)

    def remove(self, regulatory_process):
        self.children.remove(regulatory_process)

    def execute(self):
        print(f"Executing composite process: {self.process_name}")
        for child in self.children:
            child.execute()


# Leaf objects - individual steps
clinical_trial_step = Step("Clinical Trial Approval")
safety_step = Step("Safety Review")
manufacturing_step = Step("Manufacturing Validation")

# Composite processes
clinical_process = CompositeProcess("Clinical Process")
clinical_process.add(clinical_trial_step)

regulatory_process = CompositeProcess("Regulatory Process")
regulatory_process.add(safety_step)
regulatory_process.add(manufacturing_step)

full_approval_process = CompositeProcess("Full Drug Approval Process")
full_approval_process.add(clinical_process)
full_approval_process.add(regulatory_process)

full_approval_process.execute()

File: Pharmaceutical_Company_Example/Structural/JavaScript/Composite.js
// Component (Abstract class)
class RegulatoryProcess {
    execute() {
        throw new Error("This method should be implemented by subclasses.");
    }
}

// Leaf class - Step
class Step extends RegulatoryProcess {
    constructor(stepName) {
        super();
        this.stepName = stepName;
    }

    execute() {
        console.log(`Executing step: ${this.stepName}`);
    }
}

// Composite class - CompositeProcess
class CompositeProcess extends RegulatoryProcess {
    constructor(processName) {
        super();
        this.processName = processName;
        this.children = [];
    }

    add(regulatoryProcess) {
        this.children.push(regulatoryProcess);
    }

    remove(regulatoryProcess) {
        const index = this.children.indexOf(regulatoryProcess);
        if (index > -1) {
            this.children.splice(index, 1);
        }
    }

    execute() {
        console.log(`Executing composite process: ${this.processName}`);
        for (let child of this.children) {
            child.execute();
        }
    }
}

// Leaf objects - individual steps
const clinicalTrialStep = new Step("Clinical Trial Approval");
const safetyStep = new Step("Safety Review");
const manufacturingStep = new Step("Manufacturing Validation");

// Composite processes
const clinicalProcess = new CompositeProcess("Clinical Process");
clinicalProcess.add(clinicalTrialStep);

const regulatoryProcess = new CompositeProcess("Regulatory Process");
regulatoryProcess.add(safetyStep);
regulatoryProcess.add(manufacturingStep);

const fullApprovalProcess = new CompositeProcess("Full Drug Approval Process");
fullApprovalProcess.add(clinicalProcess);
fullApprovalProcess.add(regulatoryProcess);

// Execute the full approval process
fullApprovalProcess.execute();
